### PR TITLE
Add mHC pipeline and virtual-pipeline support

### PR DESCRIPTION
**Pull request by jingqiny-99**
## What does this PR do?

Enable mHC with pipeline parallelism and interleaved virtual pipelines on main. Intermediate stages carry `hidden_size * mhc_num_residual_streams` activations; both schedules now allocate their P2P buffers with that width, including physical-rank wraparound between virtual chunks.

Related to #2919. This ports the pipeline capability from #2943 onto main's mHC interfaces. The separate CUDA graph PR #7337 covers graph execution.

Tracking PR #5795 also lists this capability as P3. The earlier stacked draft #5943 was closed to rebuild P3 on main after P2; its [closure comment](https://github.com/NVIDIA/Megatron-LM/pull/5943#issuecomment-5059183514) explains that sequencing. This PR adapts the transport to main's current mHC interfaces.

The generic PP2 batched P2P message-order prerequisite overlaps open draft #7177 ([its standalone fix](https://github.com/NVIDIA/Megatron-LM/commit/9915a24149eb3f741cfc83dd624c545c71aa40ef)). Current main and dev both retain the old order, so this prerequisite is an independently reproduced pipeline bug fix. Reconcile it with #7177 as that work lands.

The deterministic GDN prerequisite addresses the same missing Q/K-normalization coverage as open dev PR #6613. This implementation also preserves FLA's backward formula using the output rounded to the input dtype, verified against the actual kernel. Main PR #6966 addresses shared autotune-cache configuration more broadly. Main PR #5396 changes the normalization call site behind an optional fusion flag; these changes should be reconciled when combined.

### Changes

- Share the pipeline residual-width calculation between the ordinary and interleaved schedules, preserving sequence/context parallel sequence dimensions and variable-length shape exchange.
- Return independent output storage from empty intermediate Hybrid stages so pipeline output deallocation preserves backward propagation.
- Keep embedding-only mHC output viewless after stream expansion, and match forward/backward batched P2P payload and shape-message ordering when PP2 shares the same peer in both directions.
- Keep Hybrid MTP on the final output chunk and reject the unsupported combination of mHC pipeline parallelism with EP communication overlap.
- Preserve existing configuration names, checkpoint keys, GPT mean contraction, Hybrid learned contraction, and eager selective mHC recomputation.
- Fix generic distributed-optimizer checkpoint prerequisites exposed by mixed FP32/BF16 mHC training: align per-parameter lookup indices with existing main-parameter order and exclude trailing DP alignment padding from unpadded checkpoint tensors. Checkpoint format remains unchanged.
- Make GDN/GDN2 Q/K normalization deterministic when requested. The native normalization preserves additive epsilon, FP32 intermediates and FLA's rounded-output backward contract; ordinary training retains the optimized FLA path.

### Validation

- **DSv4-width transport validation after merging main `0ac6ffd385` (2026-09-18):** added six real NCCL schedule tests for hidden sizes 4096 / 7168 with four mHC streams (wire widths 16384 / 28672), covering PP2, PP2/VPP2, and PP2/VPP2/TP2 + sequence parallel. On one OCI node with four GB200 GPUs and the main 26.08 ARM CI image, all four ranks reported **6 passed**; native runner exit was 0. An independent serial PyTorch reference checks every microbatch boundary, input gradient, local parameter gradient, and final loss with `rtol=0, atol=0`; the tests also verify output pseudo-deallocation. Executed at `c7cf58bcf`; the subsequent `83a5e6853` commit changes only copyright headers. These are transport-shape tests using small affine stages, not full DSv4 attention/MoE or TP/DDP gradient-finalization validation. Full DSv4 proxy testing on Lyris remains pending.
- Added real schedule comparisons against identically initialized PP1 models: loss and every finalized parameter gradient, PP2/VPP2, standalone embedding/loss chunks, empty Hybrid chunks, TP/SP + final-stage MTP, and CP with changing sequence lengths.
- Added two functional recipes for TP2/PP2/MTP and PP2/VPP2 with an empty stage and MTP.
- Main autoformatter's required Black 26.3.0, isort, pylint and Ruff checks pass; copyright and diff checks pass. Its advisory mypy step reports unresolved local GPU dependencies and existing typing errors.
- Exact main 26.08 CI image on eight H100 GPUs: **the complete 83-case PP compatibility file passed** at `eaa79ac5d`, including PP2/VPP2, embedding/loss-only and empty Hybrid chunks, TP2/SP/MTP, and all four BF16 CP/variable-length eager/recompute cases. Native runner exit and positive test counts were verified. Related Hybrid mHC and recompute files also passed 15 and 16 cases, respectively.
- The tests use native FP32 gradient accumulation and CP token-count loss normalization. Independent parity cases reset compiler caches to prevent earlier FP32 cases exhausting Dynamo's default specialization limit; the PP1/PP2 pair shares the same cache and unchanged numerical tolerances.
- Both native functional recipes complete 20 training iterations and checkpoint save after the optimizer fixes. Six CUDA checkpoint regressions pass, including mixed-dtype state ownership and actual ShardedTensor global bounds.
- GDN normalization: 8 eager/compiled FP32/BF16 kernel parity and exact-repeat cases passed; 12 full GDN/GDN2 deterministic cases passed, including TP2/SP/CP2. Fixed-input kernel characterization and independent optimizer-phase traces localized the original functional drift to this normalization path.
- **Both native functional recipes passed independent exact golden verification** on the original main 26.08 H100 CI image: TP2/PP2/MTP and TP1/PP2/VPP2 with an empty stage and MTP. Each completed all 20 iterations and checkpoint save, with native exit 0, positive verification counts, and unchanged loss, zero-gradient and memory checks. Included goldens are the untouched native output. No tolerance or required metric was relaxed, and no throughput improvement is claimed.

Local functional generation and independent verification each used `N_REPEAT=1`; the registered CI jobs request two repeats. The full registered repeat-count CI run is still pending.

The implementation is based directly on main and does not depend on the CUDA graph PR. This PR remains draft for upstream review and CI.

### CODE REVIEW

File: megatron/core/transformer/transformer_config.py [hunk 1]
```diff
@@ -2412,16 +2412,10 @@ def __post_init__(self):
         # (wrapping MoE as a HybridStack layer via HyperConnectionHybridLayer).
 
         if self.enable_mhc_connections:
-            # TransformerBlock expands to n-stream at `pre_process` and contracts back at
-            # the stage holding the final layernorm, so every intermediate pipeline stage
-            # exchanges [s, b, n*C] while the p2p buffers are still sized from hidden_size.
-            # Pipeline support must resize the p2p buffers before this guard can be lifted.
-            if self.pipeline_model_parallel_size > 1:
+            if self.pipeline_model_parallel_size > 1 and self.overlap_moe_expert_parallel_comm:
```
Comment (guihong-nv) [inline megatron/core/transformer/transformer_config.py:2415]: Can we update the inference receive buffers too, or reject mHC + PP in the inference wrapper for now? Removing this guard lets a PP2 mHC GPT model reach GPTInferenceWrapper, but AbstractModelInferenceWrapper._allocate_recv_buffer still uses hidden_size. With hidden_size=64 and two residual streams, the first stage sends width 128 while the next stage allocates width 64, so the P2P message sizes don't match. Please add a PP inference test if this path is meant to be supported.

**Review by guihong-nv — COMMENTED** [review]
I found one remaining shape mismatch in the inference path. I checked the production shape calculations, but couldn't run the GPU suite locally.

File: megatron/core/transformer/transformer_config.py (hunk 1 above)
Comment (FDecaYed) [reply megatron/core/transformer/transformer_config.py:2415]: Fixed in 52f6f83ebe: the inference wrapper now rejects mHC + PP based on the inference process group. Added a focused regression test; training PP/VPP remains enabled.
